File: Python/linear_structures/linked_list.py
'''linked list'''


class Node(object):
    '''Node class for a node of a linked list'''

    def __init__(self, data) -> None:
        self.data = data
        self.next = None


class LinkedList():
    '''Linked list class'''

    def __init__(self) -> None:
        self.head = None

    def __len__(self) -> int:
        current_node = self.head
        count = 0
        while current_node:
            count += 1
            current_node = current_node.next
        return count

    def __str__(self) -> str:
        string = 'Head '

        if self.is_empty():
            string += '-> None'
            return string

        current_node = self.head
        while current_node:
            string += f'-> {current_node.data} '
            current_node = current_node.next

        return string

    def add(self, item: any) -> None:
        '''Add and item to the linked list
        Note: adds node to the beggining or the head of the linked list.'''
        temp = Node(item)
        temp.next = self.head
        self.head = temp

    def remove(self, item: any) -> None:
        '''Finds and removes an item from list'''
        found = False
        current = self.head
        previous = None
        while not found and current:
            if current.data == item:
                found = True
            else:
                previous = current
                current = current.next
        if not found:
            raise IndexError("Item not found.")
        if previous is None:
            self.head = current.next
        else:
            previous.next = current.next

    def pop(self) -> None:
        '''Pops first item from linked list'''
        if self.is_empty():
            raise IndexError("Can't pop from empty list.")

        temp = self.head.next
        self.head = temp

    def contains(self, item: any) -> bool:
        '''Checks if the item is in the linked list and returns true if it is false otherwise'''
        found = False
        current_node = self.head
        while not found and current_node:
            if current_node.data == item:
                found = True
            current_node = current_node.next
        return found

    def is_empty(self) -> bool:
        '''returns true if the list is empty'''
        return self.head is None

File: C/linear_structures/linked_list.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node
{
    int value;
    struct node *next;
} node_t;

void insert_node(int value);
void delete_node();
void print_linked_list(node_t *head);

node_t *head = NULL;

int main()
{
    print_linked_list(head);

    insert_node(7);

    insert_node(9);

    insert_node(4);
    print_linked_list(head);
    delete_node();
    delete_node();
    print_linked_list(head);
}

// inserts node at the front of the linked list
void insert_node(int value)
{
    node_t *new = (node_t *)malloc(sizeof(node_t));

    if (new == NULL)
    {
        printf("\nCannot create the Node");
        return;
    }

    new->value = value;
    new->next = head;
    head = new;
}

// inserts node from the front of the linked list
void delete_node()
{
    if (head == NULL)
    {
        printf("\nlist is empty");
        return;
    }
    node_t *temp = head;
    head = head->next;
    free(temp);
}

void print_linked_list(node_t *head)
{

    node_t *current = head;
    printf("\nstart -> ");
    while (current != NULL)
    {
        printf("%d -> ", current->value);
        current = current->next;
    }
    printf("NULL \n");
}